### PR TITLE
Add / switch to Go 1.18

### PR DESCRIPTION
**Pull request by kolyshkin**
- switch to golangci-lint 1.45 (which is go1.18 compatible);
 - switch to Go 1.18 as the default Go version in CI;
 - remove 1.16 compatibility, and remove 1.16 from CI matrix.

### CODE REVIEW

**Comment by kolyshkin** [comment]
Ah! golangci-lint is not working with go 1.18 yet (see https://github.com/golangci/golangci-lint/pull/2438).

**Comment by kolyshkin** [comment]
`test (1.17.x, criu-dev)` is not shown in checks since it became `test (1.18.x, criu-dev)`.

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -21,13 +21,13 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        go-version: [1.16.x, 1.17.x, 1.18.0-beta1]
+        go-version: [1.16.x, 1.17.x, 1.18.x]
```
Comment (AkihiroSuda) [inline .github/workflows/test.yml:24]: We can drop 1.16

**Review by AkihiroSuda — DISMISSED** [review]

File: .github/workflows/validate.yml [hunk 2]
```diff
@@ -18,7 +18,7 @@ jobs:
       - uses: actions/checkout@v3
       - uses: actions/setup-go@v2
         with:
-          go-version: "${{ env.GO_VERSION }}"
+          go-version: 1.17.x # TODO switch to "${{ env.GO_VERSION }}"
```
Comment (thaJeztah) [inline .github/workflows/validate.yml:21]: Looks like go 1.18 may be supported now (although it looks like they will skip some linters, based on some tickets linked to that one); not sure if we're running all of those though; https://github.com/golangci/golangci-lint/pull/2438

Comment (kolyshkin) [reply .github/workflows/validate.yml:21]: Let me try again )

Comment (thaJeztah) [reply .github/workflows/validate.yml:21]: Yeah, no promises; just noticed the linked ticket was closed, so perhaps we could have everything done in one go 👍 🤞

File: .github/workflows/test.yml (hunk 1 above)
Comment (kolyshkin) [reply .github/workflows/test.yml]: Good; modified accordingly. The change to drop 1.16 is somewhat big so implemented in a separate commit.

**Comment by kolyshkin** [comment]
> Looking better; one deprecation warning about `strings.Title()`;
> 
> ```
>   Error: SA1019: strings.Title is deprecated: The rule Title uses for word boundaries does not handle Unicode punctuation properly. Use golang.org/x/text/cases instead.  (staticcheck)
> ```
> 
> My recommendation would be to add a `nolint` for that one, as it's perfectly fine to use for ASCII(ish) characters; their recommendation is ridiculous, and brings in 20K lines of code (see [moby/moby#43201](https://github.com/moby/moby/pull/43201))

I ended up with own implementation of `title()` which is just enough for the purposes.

If a function is deprecated, it might be removed later, so I'd rather not use it.

**Comment by thaJeztah** [comment]
> If a function is deprecated, it might be removed later, so I'd rather not use it.

It's in stdlib, so won't be removed until Go 2.0 at least, but it's silly they marked it as deprecated. The discussion that lead to that was because there was a "note" in the function's description that there were some corner cases where it didn't work, then someone suggested; let's then deprecate it!

Instead, they could've just documented its limitations. The same problem applies to most of stdlib (no full support for all unicode corner cases).

**Comment by kolyshkin** [comment]
OK, Fedora 35 comes with Go 1.16 so we need to install golang if we drop Go 1.16 support.

**Review by dqminh — APPROVED** [review]

File: libcontainer/container_linux.go [hunk 3]
```diff
@@ -919,8 +919,19 @@ func (c *linuxContainer) criuSupportsExtNS(t configs.NamespaceType) bool {
...
+		return s
+	}
+	s0 -= 'a' - 'A' // convert to uppercase
+
+	return string(s0) + s[1:]
+}
+
 func criuNsToKey(t configs.NamespaceType) string {
-	return "extRoot" + strings.Title(configs.NsName(t)) + "NS"
+	return "extRoot" + title(configs.NsName(t)) + "NS"
```
Comment (AkihiroSuda) [inline libcontainer/container_linux.go:934]: Is this parsed by something?
Do we really need to titlize it?

**Review by AkihiroSuda — APPROVED** [review]

File: libcontainer/container_linux.go [hunk 4]
```diff
@@ -919,8 +919,19 @@ func (c *linuxContainer) criuSupportsExtNS(t configs.NamespaceType) bool {
 	return c.checkCriuVersion(minVersion) == nil
 }
 
+// title converts the first character of ASCII string s to uppercase.
+func title(s string) string {
+	s0 := s[0]
```
Comment (thaJeztah) [inline libcontainer/container_linux.go:924]: This will panic if string would ever be empty; is that something that could possibly happen? (also could we get non-ascii (multi-byte) characters, in which case we'd be doing the wrong thing?)

File: libcontainer/container_linux.go (hunk 3 above)
Comment (kolyshkin) [reply libcontainer/container_linux.go:934]: Quoting commit message

> The alternative is to not capitalize the namespace string; this will break
> restoring of a container checkpointed by earlier version of runc.

File: libcontainer/container_linux.go (hunk 4 above)
Comment (kolyshkin) [reply libcontainer/container_linux.go:924]: If the first character is non-ASCI (e.g. хлеб), it will be returned as-is (no capitalization).

Indeed it will panic with empty string as an argument, and it is theoretically possible to have it.

I guess I'll just go back to using strings.Title and suppressing the deprecation warning.

File: libcontainer/container_linux.go (hunk 3 above)
Comment (kolyshkin) [reply libcontainer/container_linux.go]: I moved back to using `strings.Title` instead, suppressing the deprecation warning as it seems easier.

**Comment by kolyshkin** [comment]
I have switched back to using `strings.Title` (suppressing the deprecation warning as it was suggested by @thaJeztah earlier). Not other changes, please re-LGTM @thaJeztah @AkihiroSuda @dqminh.

I will merge this myself since this involves changing the branch protection rules for `main` (remove `test (1.16.x, ...)`, add `test(1.18.x, ...)`, change `test (1.17.x, criu-dev)` to 1.18.x).

File: libcontainer/container_linux.go (hunk 4 above)
Comment (thaJeztah) [reply libcontainer/container_linux.go]: Ah, yeah, perhaps best to do that.

I'm considering proposing to un-deprecate this function (I think it's more disruptive than useful)

^^ feel free to beat me to that though if you agree 😁

**Review by thaJeztah — APPROVED** [review]
LGTM

**Review by AkihiroSuda — APPROVED** [review]

**Comment by kolyshkin** [comment]
Branch protection rules updated. This also means all older PRs need to be re-tested...
